src/libValidator/Validator.cpp

```diff
@@ -350,3 +350,59 @@ bool Validator::CheckDirBlocks(
   newDSComm = move(mutable_ds_comm);
   return ret;
 }
+
+ValidatorBase::TxBlockValidationMsg Validator::CheckTxBlocks(
+    const vector<TxBlock>& txBlocks, const deque<pair<PubKey, Peer>>& dsComm,
+    const BlockLink& latestBlockLink) {
+  // Verify the last Tx Block
+  uint64_t latestDSIndex = get<BlockLinkIndex::DSINDEX>(latestBlockLink);
+
+  if (get<BlockLinkIndex::BLOCKTYPE>(latestBlockLink) != BlockType::DS) {
+    if (latestDSIndex == 0) {
+      LOG_GENERAL(WARNING, "The latestDSIndex is 0 and blocktype not DS");
+      return TxBlockValidationMsg::INVALID;
+    }
+    latestDSIndex--;
+  }
+
+  const TxBlock& latestTxBlock = txBlocks.back();
+
+  if (latestTxBlock.GetHeader().GetDSBlockNum() != latestDSIndex) {
+    if (latestDSIndex > latestTxBlock.GetHeader().GetDSBlockNum()) {
+      LOG_GENERAL(WARNING, "Latest Tx Block fetched is stale ");
+      return TxBlockValidationMsg::INVALID;
+    }
+
+    LOG_GENERAL(WARNING,
+                "The latest DS index does not match that of the latest tx "
+                "block ds num, try fetching Tx and Dir Blocks again "
+                    << latestTxBlock.GetHeader().GetDSBlockNum() << " "
+                    << latestDSIndex);
+    return TxBlockValidationMsg::STALEDSINFO;
+  }
+
+  if (!CheckBlockCosignature(latestTxBlock, dsComm)) {
+    return TxBlockValidationMsg::INVALID;
+  }
+
+  if (txBlocks.size() < 2) {
+    return TxBlockValidationMsg::VALID;
+  }
+
+  BlockHash prevBlockHash = latestTxBlock.GetHeader().GetPrevHash();
+  unsigned int sIndex = txBlocks.size() - 2;
+
+  for (unsigned int i = 0; i < txBlocks.size() - 1; i++) {
```

Comment:
use range based loop?

Reply:
need to run to size() - 1

Reply:
OK

Reply:
on further investigation, as I use sIndex to fetch an address, this cannot be changed in to range-based.